include/wlr/types/wlr_layer_shell.h

```diff
@@ -0,0 +1,102 @@
+#ifndef WLR_TYPES_WLR_LAYER_SHELL_H
+#define WLR_TYPES_WLR_LAYER_SHELL_H
+#include <stdbool.h>
+#include <stdint.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_box.h>
+#include <wlr/types/wlr_surface.h>
+#include "wlr-layer-shell-unstable-v1-protocol.h"
+
+/**
+ * wlr_layer_shell allows clients to arrange themselves in "layers" on the
+ * desktop in accordance with the wlr-layer-shell protocol. When a client is
+ * added, the new_surface signal will be raised and passed a reference to our
+ * wlr_layer_surface. At this time, the client will have configured the surface
+ * as it desires, including information like desired anchors and margins. The
+ * compositor should use this information to decide how to arrange the layer
+ * on-screen, then determine the dimensions of the layer and call
+ * wlr_layer_surface_configure. The client will then attach a buffer and commit
+ * the surface, at which point the wlr_layer_surface map signal is raised and
+ * the compositor should begin rendering the surface.
+ */
+struct wlr_layer_shell {
+	struct wl_global *wl_global;
+	struct wl_list clients;
+
+	struct wl_listener display_destroy;
+
+	struct {
+		struct wl_signal new_surface;
+	} events;
+
+	void *data;
+};
+
+struct wlr_layer_client {
+	struct wlr_layer_shell *shell;
+	struct wl_resource *resource;
+	struct wl_client *client;
+	struct wl_list surfaces;
+
+	struct wl_list link; // wlr_layer_shell::clients
+};
+
+struct wlr_layer_surface_state {
+	// Client
+	uint32_t anchor;
+	uint32_t exclusive_zone;
+	struct {
+		uint32_t top, right, bottom, left;
+	} margin;
+	// Server
+	uint32_t width, height;
+};
+
+struct wlr_layer_surface_configure {
+	struct wl_list link; // wlr_layer_surface::configure_list
+	uint32_t serial;
+	struct wlr_layer_surface_state state;
+};
+
+struct wlr_layer_surface {
+	struct wlr_surface *surface;
+	struct wlr_layer_client *client;
+	struct wl_resource *resource;
+	struct wl_list link; // wlr_layer_client:surfaces
+
+	const char *namespace;
+	enum zwlr_layer_shell_v1_layer layer;
+
+	bool added, configured, mapped;
+	uint32_t configure_serial;
+	struct wl_event_source *configure_idle;
+	uint32_t configure_next_serial;
+	struct wl_list configure_list;
+
+	struct wlr_layer_surface_state next; // client protocol requests
+	struct wlr_layer_surface_state pending; // our configure requests
```

Comment:
Can we think of better names for next/pending that reflects these comments?

Reply:
Yes